rfcs/framework-0072-pluggable-transfers.md

```diff
@@ -0,0 +1,507 @@
+We propose integrating a custom file uploader UI based on the Uppy.io library, utilizing the AWS S3 Multipart uploader
+plugin.
+
+## Motivation
+
+We plan to introduce a built-in **Multipart** transfer type that enables
+uploading large files in separate chunks, improving performance and
+reliability.
+
+In addition, one could, for example, develop an **S3Copy** transfer type to
+copy files directly between S3 buckets—bypassing the Invenio server—to reduce
+bandwidth usageand improve efficiency.
+
+As a record depositor, I want to:
+- Select one or multiple local files to upload and assign to a record.
+- Select one or multiple remote files (e.g., import from a parent record or cloud storage services).
+- View the upload progress of each ongoing file upload.
+- Abort an ongoing upload.
+- Retry any failed uploads.
+- View and manage already uploaded files (e.g., remove files, update associated metadata such as image captions and descriptions).
+- Verify file integrity using checksums when uploading (to ensure the file was received and stored without errors) or downloading (to confirm I received an exact copy as stored on the server).
+- Efficiently and reliably upload large files (>10 GiB) even over unreliable networks.
+
+As a developer, I want to:
+- Implement or integrate custom file uploader UIs (including custom upload logic) into the deposit form (e.g., using library like Uppy.io), allowing the front-end to support any configured transfer types.
+
+As an admin, I want to:
+- Configure the default upload mechanism (TransferType) for new uploads.
+- Configure which uploader UI should be used in the deposit form.
+- Define upload restrictions (e.g., allowed MIME types, file size limits).
+
+## REST API changes
+
+To make file transfers more consistent, we are introducing a new `transfer`
+field to replace the current inconsistent usage of `storage_class`.
+
+- In **invenio-files-rest**, `storage_class` indicates the file’s intended location or usage (e.g., `archive`, `standard`).  
+- In **invenio-records-resources**, it has been (mis)used to denote the transfer type (e.g., `local`, `fetch`, `remote`).
+
+Going forward, the new `transfer` field will hold all transfer-related
+information, while `storage_class` will continue to serve its original purpose
+from `invenio-files-rest`.
+
+The `transfer` field will be a dictionary with the following structure:
+
+### Local transfer
+
+Equivalent to current `storage_class == "L"`. The `transfer` section can be omitted.
+
+```json
+[{
+    "key": "file.txt",
+    "transfer": {
+      "type": "L",
+    }
+}]
+```
+
+### Fetch transfer
+
+Equivalent to current storage_class == "F"
+
+```json
+[{
+    "key": "file.txt",
+    "transfer": {
+        "type": "F",
+        "url": "https://example.com/file-to-fetch.txt",
+    }
+}]
+```
+
+### Remote transfer
+
+Currently not supported on REST API level.
+
+```json
+[{
+    "key": "file.txt",
+    "transfer": {
+        "type": "R",
```

Comment:
I hope the amount of types won't explode in the future, causing naming clashes :)